README
Labyrinth-Demo
==============

Documentation overview for the Demo Website of the Labyrinth Website Management
Framework.

Labyrinth began life in 2002, with a small set of plugins to enable various 
features of web site management. This demo website uses the core set of plugins 
to provide an example of a website management system.

These plugins provide the functionality to manipulate core aspects of the data 
used within Labyrinth. Several of these plugins only provide administration 
features.


ADDITIONAL FILES
----------------

In order for your website to work with Labyrinth, some additional files are 
required. This package provides the standard set of files to enable Labyrinth 
and the Core plugins to work with a blank canvas. These files consist of SQL, 
template and configuration files, together with some basic CSS and Javascript 
files.

Install these into the website structure and edit as necessary to implement 
the layout and functionality you require for your purposes.

Please note that included with this distribution are files designed and 
developed by other people. Notable are the following:

   PixelGreen
       Version 1.2

       CSS Layout and Design by Erwin Aligam, release under the Creative
       Commons Attribution 2.5 License -
       http://creativecommons.org/licenses/by/2.5/

       See http://www.styleshout.com/ for more details.

   TinyMCE
       Version 3.211

       A Javascript WYSIWYG Content Editor by Moxiecode Systems AB, released
       under the GNU Leseer General Public License (LGPL) -
       http://tinymce.moxiecode.com/js/tinymce/jscripts/tiny_mce/license.txt

       See http://tinymce.moxiecode.com/ for more details.


INSTALLATION
------------
For more complete installtion instructions, please see the INSTALL file.

To set up the Demo Website, install via a CPAN client to ensure you install 
all the required prerequisites. Then copy the directory contents under the 
'vhost' directory into your web server directory. For this example the path
to these files and directories is assumed to be '/var/www/demo'. If you wish 
to change this, please update the file paths used in 'cgi-bin/pages.cgi' and
'cgi-bin/config/settings.ini'

If you use Apache, include the configuration snippet contained in the file 
'vhosts.conf', in the appropriate Apache configuration file. Directions for 
other web servers will be included at a later date.

Create a MySQL database called 'demo' and use the base file
'cgi-bin/db/demo-base.sql' as a starting point. If you wish to specify a fixed
user to access the database, you may wish to review the 'grant.sql' file or 
amend the 'settings.ini' as appropriate.

You should now be able to start your Apache instance. Update your hosts file as
necessary to point to your new virtual host.


LABYRINTH DIRECTORY STRUCTURE
-----------------------------

    ./cgi-bin
        /config         - contains top level configuration files
           /requests    - contains despatch tables
        /db                     - database dumps
        /lib
           /Labyrinth   - Labyrinth Core modules
              /Plugin   - Labyrinth Plugin modules
        /templates              - Template Toolkit templates
    ./html
        /cache          - cache where reports are produced
        /css            - website CSS files
        /images         - website image files
        /js                     - website JavaScript files
           /tiny_mce    - Tiny MCE JavaScript application files
    ./toolkit

As mentioned included in the package is the vhost entry from an Apache
configuration file. This file contains the rewrite rules that translate
user friendly URLs to the local CGI call. However, it is not necessary
to have rewrite rules, in the event you are unable to use them, and the
application can quite happily function with full 'cgi-bin/' type paths.

The './cgi-bin/pages.cgi' file is the core startup script, that simply
passes the settings file, and initiates Labyrinth. The 'act' CGI
parameter contains the primary action required. Throughout the duration
of a request, the current action may change several times, with each
referencing the despatch tables to load and run the correct functions
and render the correct templates. Each user has a nominated "realm",
which is the section of the site they are given access to. The realm
despatch table is used to load and run and default functions, and can
thus be used to render different skins and/or layout templates.

The CGI parameters are validated against the
'./cgi-bin/config/parserules.ini' file. The rules are built from these
entries and passed to Data::FormValidator.  Each entry's fields refer
to parameter name, whether mandatory, default value, preprocessing
function, constraint function and finally the regular expression.  If a
constraint is provided the regular expression is ignored. Where the
parameter name looks like ':^LISTED', this is a regular expression
match against the parameter name (where the ':' character is removed).
This enables parameters which have a changeable suffix, can apply the
same constraint or regular expression against each matching parameter's
value.

All SQL statements are called using a Phrasebook Design Pattern. As
such the './cgi-bin/config/phrasebook.ini' file contains all each
phrase and associated SQL as used within the application. To aid
readability SQL can be split onto multiple lines, using the '\'
character as the last entry on a line to continue onto the next.

Templates are built from the given layout, see the appropriate realm
entry in the despatch table, incorporating other templates as required,
including the 'content' template as specified by the last action
command. See the appropriate despatch table to reference the
appropriate content template.

Tiny MCE is used by Labyrinth where appropriate for text area input.
However, this is only activated on the pages where it is appropriate.

The set of scripts within the toolkit directory typically provide ad-
hoc or timed (e.g. via cron) functionality. The scripts are not
intended to be run via the web server, but often provide support
functionality, such as creating static pages or running periodic
cleanup tasks.


DEMO WEBSITE
------------

To see a working example of this distribution, please see the following link:

  <http://demo.missbarbell.co.uk>


ADDITION INFORMATION
--------------------

Although Labyrinth has long been in development, documentation has not been a 
priority. As such much of the documentation you may need to understand how to
use Labyrinth is the code itself. If you have the inclination, documentation 
patches would be very gratefully received.

The Labyrinth website [1] will eventually feature a documentation site, wiki 
and other features which are intended to provide you with the information to 
use and extend Labyrinth as you wish.

  [1] http://labyrinth.missbarbell.co.uk


SEE ALSO
--------

  Labyrinth, Labyrinth-Plugin-Core

  http://labyrinth.missbarbell.co.uk


AUTHOR
------

  Barbie, <barbie@missbarbell.co.uk> for Miss Barbell Productions,
  <http://www.missbarbell.co.uk/>


COPYRIGHT & LICENSE
-------------------

  Copyright (C) 2002-2014 Barbie for Miss Barbell Productions
  All Rights Reserved.

  This module is free software; you can redistribute it and/or
  modify it under the Artistic License 2.0.


Changes
Changes for the Labyrinth Demo Site
===================================

        - fixed license fields in META.json to be lists.
        - extended META test suite.
        - upgrade to Labyrinth-5.29.
        - upgrade to Labyrinth-Plugin-Core-5.18.
        - database table charset changed to UTF8.
        - phasebook updates.

1.03    2014-08-16
        - fixed distribution name in metadata.

1.02    2014-02-02
        - removed IMAGE parse rule.
        - updated templates to XHTML standards.
        - upgrade to Labyrinth-5.20.
        - upgrade to Labyrinth-Plugin-Core-5.14.
        - removed obsolete templates & images.

1.01    2013-02-10
        - updates to phrasebook.ini.
        - added indexes to SQL.
        - change file dates changed to meet W3CDTF standards.
        - increased Labyrinth Core version dependency.
        - updated and added tinymce.
        - reworked Makefile.PL for clarity.
        - added LICENSE and INSTALL files.

1.00    2011-01-01
        - official release alongside Labyrinth Core.


vhost/toolkit/sumhits.pl
#!/usr/bin/perl -w
use strict;

$|++;

=head1 NAME

sumhits - Sum all hit counters, older than a month

=head1 SYNOPSIS

  perl sumhits.pl

=head1 DESCRIPTION

This application retrieves all the hit counters older than one month, adds the
same counters and updates the database with a single entry total for each.

=cut

my $BASE;

# create images in pages
BEGIN {
    $BASE = '/var/www/demo';
}

#----------------------------------------------------------
# Additional Modules

use lib qw|../cgi-bin/lib|;
use Labyrinth::DBUtils;
use Labyrinth::Globals;
use Labyrinth::Variables;

use Getopt::Long;

#----------------------------------------------------------
# Code

Labyrinth::Globals::LoadSettings("$BASE/toolkit/sumhits-settings.ini");
$settings{autocommit} = 0;
Labyrinth::Globals::DBConnect();

my %options;
GetOptions( \%options, 'datetime|d=i' ) or die;
my $datetime = $options{datetime} ? $options{datetime} : time() - 60 * 60 * 24 * 7 * 6;

my @areas = $dbi->GetQuery('hash','GetHitAreas');
for my $rs (@areas) {
    my $area = $rs->{area};
    my @rows = $dbi->GetQuery('hash','SumHits',$datetime,$area);
    for my $row (@rows) {
        # no processing needed?
        next    if($row->{number} == 1 && $row->{createdate} == 0);

        $dbi->DoQuery('StartTrans');
        my $rows = $dbi->DoQuery('DelAHit',$datetime,$row->{area},$row->{pageid},$row->{photoid},$row->{query});
        $dbi->DoQuery('AddAHit',$row->{counter},$row->{area},$row->{pageid},$row->{photoid},$row->{query},0);
        $dbi->DoQuery('CommitTrans');
        printf "[%6d] %6d,%15s,%3d,%3d,%s\n", $rows, $row->{counter},$row->{area},$row->{pageid},$row->{photoid},$row->{query};
    }
}

__END__

=head1 AUTHOR

  Copyright (c) 2002-2011 Barbie <barbie@cpan.org> Miss Barbell Productions.

=head1 LICENSE

  This program is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself.

  See http://www.perl.com/perl/misc/Artistic.html

=cut


vhost/toolkit/rssfeeds.pl
#!/usr/bin/perl -w
use strict;

=head1 NAME

rssfeeds - Provide the RSS feeds for the site

=head1 SYNOPSIS

  perl rssfeeds.pl

=head1 DESCRIPTION

This application provides all the RSS feeds for the website.

=cut

my $BASE;

# create images in pages
BEGIN {
    $BASE = '/var/www/demo';
}

#----------------------------------------------------------
# Additional Modules

use lib qw|../cgi-bin/lib|;
use Labyrinth::Globals;
use Labyrinth::RSS;
use Labyrinth::Variables;
use Labyrinth::Plugin::Articles;

use IO::File;

#----------------------------------------------------------
# Variables

my @types = (
#    { type => 'rss',  version => '0.9' },
#    { type => 'rss',  version => '1.0' },
    { type => 'rss',  version => '2.0' },
    { type => 'atom', version => '1.0' },
);

#----------------------------------------------------------
# Code

Labyrinth::Globals::LoadSettings("$BASE/cgi-bin/config/settings.ini");
Labyrinth::Globals::DBConnect();

# Diary Entries
$settings{perma} = $tvars{webpath} . '/diary/';
$cgiparams{sectionid} = 6;
$settings{limit} = 10;

my $arts = Labyrinth::Plugin::Articles->new();
$arts->List();


for my $item (@types) {
    my $rss = Labyrinth::RSS->new( %$item, perma => 'http://demo.example.com/' );
    my $xml = $rss->feed(@{$tvars{articles}});
    write_xml("rss/$item->{type}-$item->{version}.xml",$xml);
}

sub write_xml {
    my $file = shift;
    my $xml  = shift;
    my $fh = IO::File->new("$BASE/html/$file",'w')    or die "Cannot write to file [$BASE/html/$file]: $!";
    print $fh $xml;
    $fh->close;
}

__END__

=head1 AUTHOR

  Copyright (c) 2002-2011 Barbie <barbie@cpan.org> Miss Barbell Productions.

=head1 LICENSE

  This program is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself.

  See http://www.perl.com/perl/misc/Artistic.html

=cut


lib/Labyrinth/Demo.pm
package Labyrinth::Demo;

use warnings;
use strict;

our $VERSION = '1.03';

1;

__END__

=head1 NAME

Labyrinth::Demo - Labyrinth Website Management Framework - Demo Website

=head1 DESCRIPTION

Documentation overview for the Demo Website of the Labyrinth Website 
Management Framework.

Labyrinth began life in 2002, with a small set of plugins to enable various
features of web site management. 

This demo website uses the core set of plugins to provide an example of a
website management system. 

These plugins provide the functionality to manipulate core aspects of the data
used within Labyrinth. Several of these plugins only provide administration 
features.

=head1 ADDITIONAL FILES

In order for your website to work with Labyrinth, some additional files are 
required. This package provides the standard set of files to enable Labyrinth
and the Core plugins to work with a blank canvas. These files consist of SQL, 
template and configuration files, together with some basic CSS and Javascript 
files.

Install these into the website structure and edit as necessary to implement
the layout and functionality you require for your purposes.

Please note that included with this distribution are files designed and 
developed by other people. Notable are the following:

=head2 PixelGreen

Version 1.2

CSS Layout and Design by Erwin Aligam, release under the Creative Commons 
Attribution 2.5 License - http://creativecommons.org/licenses/by/2.5/

See http://www.styleshout.com/ for more details.

=head2 TinyMCE

Version 3.211

A Javascript WYSIWYG Content Editor by Moxiecode Systems AB, released under the
GNU Leseer General Public License (LGPL) - 
http://tinymce.moxiecode.com/js/tinymce/jscripts/tiny_mce/license.txt

See http://tinymce.moxiecode.com/ for more details.

=head1 INSTALLATION

To set up the Demo Website, install via a CPAN client (such as CPAN.pm, 
CPANPLUS or cpanminus) to ensure you install required prerequisites. Then copy
the directory contents under the 'vhost' directory into your web server 
directory. For this example the path to these files and directories is assumed
to be '/var/www/demo'. If you wish to change this, please update the file paths
used in 'cgi-bin/pages.cgi' and 'cgi-bin/config/settings.ini'

If you use Apache, include the configuration snippet contained in the file
'vhosts.conf', in the appropriate Apache configuration file. Directions for
other web servers will be included at a later date.

Create a MySQL database called 'demo' and use the base file 
'cgi-bin/db/demo-base.sql' as a starting point. If you wish to specify a fixed
user to access the database, you may wish to review the 'grant.sql' file or
amend the 'settings.ini' as appropriate.

You should now be able to start your Apache instance. Update your hosts file 
as necessary to point to your new virtual host.

A more complete installation and configuration script is planned for the 
future, which can also install plugins.

=head1 LABYRINTH DIRECTORY STRUCTURE

    ./cgi-bin
        /config		- contains top level configuration files
           /requests	- contains despatch tables
        /db			- database dumps
        /lib
           /Labyrinth	- Labyrinth Core modules
              /Plugin	- Labyrinth Plugin modules
        /templates		- Template Toolkit templates
    ./html
        /cache		- cache where reports are produced
        /css		- website CSS files
        /images		- website image files
        /js			- website JavaScript files
           /tiny_mce	- Tiny MCE JavaScript application files
    ./toolkit

As mentioned included in the package is the vhost entry from an Apache 
configuration file. This file contains the rewrite rules that translate user 
friendly URLs to the local CGI call. However, it is not necessary to have 
rewrite rules, in the event you are unable to use them, and the application 
can quite happily function with full 'cgi-bin/' type paths.

The './cgi-bin/pages.cgi' file is the core startup script, that simply passes
the settings file, and initiates Labyrinth. The 'act' CGI parameter contains 
the primary action required. Throughout the duration of a request, the current
action may change several times, with each referencing the despatch tables to 
load and run the correct functions and render the correct templates. Each user 
has a nominated "realm", which is the section of the site they are given access
to. The realm despatch table is used to load and run and default functions, and
can thus be used to render different skins and/or layout templates.

The CGI parameters are validated against the './cgi-bin/config/parserules.ini'
file. The rules are built from these entries and passed to Data::FormValidator. 
Each entry's fields refer to parameter name, whether mandatory, default value, 
preprocessing function, constraint function and finally the regular expression. 
If a constraint is provided the regular expression is ignored. Where the 
parameter name looks like ':^LISTED', this is a regular expression match 
against the parameter name (where the ':' character is removed). This enables 
parameters which have a changeable suffix, can apply the same constraint or 
regular expression against each matching parameter's value.

All SQL statements are called using a Phrasebook Design Pattern. As such the 
'./cgi-bin/config/phrasebook.ini' file contains all each phrase and associated 
SQL as used within the application. To aid readability SQL can be split onto 
multiple lines, using the '\' character as the last entry on a line to continue 
onto the next.

Templates are built from the given layout, see the appropriate realm entry in 
the despatch table, incorporating other templates as required, including the 
'content' template as specified by the last action command. See the appropriate
despatch table to reference the appropriate content template.

Tiny MCE is used by Labyrinth where appropriate for text area input. However, 
this is only activated on the pages where it is appropriate. 

The set of scripts within the toolkit directory typically provide ad-hoc or 
timed (e.g. via cron) functionality. The scripts are not intended to be run via
the web server, but often provide support functionality, such as creating 
static pages or running periodic cleanup tasks.

=head1 DEMO WEBSITE

To see a working example of this distribution, please see the following link:

L<http://demo.missbarbell.co.uk>

=head1 ADDITION INFORMATION

Although Labyrinth has long been in development, documentation has not been a
priority. As such much of the documentation you may need to understand how to
use Labyrinth is the code itself. If you have the inclination, documentation
patches would be very gratefully received.

The Labyrinth website [1] will eventually feature a documentation site, wiki
and other features which are intended to provide you with the information to
use and extend Labyrinth as you wish.

[1] http://labyrinth.missbarbell.co.uk

=head1 SEE ALSO

L<Labyrinth>,
L<Labyrinth-Plugin-Core>

http://labyrinth.missbarbell.co.uk

=head1 AUTHOR

Barbie, <barbie@missbarbell.co.uk> for
Miss Barbell Productions, L<http://www.missbarbell.co.uk/>

=head1 COPYRIGHT & LICENSE

  Copyright (C) 2002-2014 Barbie for Miss Barbell Productions
  All Rights Reserved.

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic License 2.0.

=cut
